In [3]:
import pandas as pd
import csv

In [4]:
df2 = pd.read_csv('Products.csv')
df2.head()

ParserError: Error tokenizing data. C error: Buffer overflow caught - possible malformed input file.


Not sure why trying to import the data from the csv file straight into a dataframe is causing errors, but a workaround is required.

In [5]:
with open('Products.csv') as file:
    csv_reader = csv.reader(file)
    header = []
    header = next(csv_reader)
    
    rows = []
    for row in csv_reader:
        rows.append(row)

In [6]:
header

['',
 'id',
 'product_name',
 'category',
 'product_description',
 'price',
 'location',
 'url',
 'page_id',
 'create_time']

In [7]:
rows[0:2]

[['0',
  '243809c0-9cfc-4486-ad12-3b7a16605ba9',
  'Mirror wall art | in Wokingham, Berkshire | Gumtree',
  'Home & Garden / Dining, Living Room Furniture / Mirrors, Clocks & Ornaments',
  'Mirror wall art. Posted by Nisha in Dining, Living Room Furniture, Mirrors, Clocks & Ornaments in Wokingham. 25 February 2022Size 45x45cm',
  '£5.00',
  'Wokingham, Berkshire',
  'https://www.gumtree.com/p/mirrors-clocks-ornaments/mirror-wall-art/1426704584',
  '1426704584',
  '2022-02-26'],
 ['1',
  '1c58d3f9-8b93-47ea-9415-204fcc2a22e6',
  'Stainless Steel Food Steamer | in Inverness, Highland | Gumtree',
  'Home & Garden / Other Household Goods',
  'Morphy Richard’s (model no 48755)Stainless steel 3 tier stackable food steamer9 litre capacity 1.1 litre rice tray Easy clean and can be used in dishwasher Ideal for family cookingServes up to 4 personBrand New & never used Still in original packaging ',
  '£20.00',
  'Inverness, Highland',
  'https://www.gumtree.com/p/other-household-goods/stainless-

In [8]:
df = pd.DataFrame(rows, columns=header)
df.head()

id  \
0  0  243809c0-9cfc-4486-ad12-3b7a16605ba9   
1  1  1c58d3f9-8b93-47ea-9415-204fcc2a22e6   
2  2  860673f1-57f6-47ba-8d2f-13f9e05b8f9a   
3  3  59948726-29be-4b35-ade5-bb2fd7331856   
4  4  16dbc860-696e-4cda-93f6-4dd4926573fb   

                                        product_name  \
0  Mirror wall art | in Wokingham, Berkshire | Gu...   
1  Stainless Steel Food Steamer | in Inverness, H...   
2  Sun loungers | in Skegness, Lincolnshire | Gum...   
3  Coffee side table from Ammunition ammo box hai...   
4  Modern Shannon Sofa for sale at low cost | in ...   

                                            category  \
0  Home & Garden / Dining, Living Room Furniture ...   
1              Home & Garden / Other Household Goods   
2  Home & Garden / Garden & Patio / Outdoor Setti...   
3  Home & Garden / Dining, Living Room Furniture ...   
4  Home & Garden / Dining, Living Room Furniture ...   

                                 product_description    price  \
0  Mirror wall art. Posted by Nisha in Dining, Li...    £5.00   
1  Morphy Richard’s (model no 48755)Stainless ste...   £20.00   
2  I have 2 of these - collection only as I don’t...   £20.00   
3  Great reclaimed army ammunition box used as co...  £115.00   
4  New Design Shannon Corner sofa  5 Seater Avail...     None   

                 location                                                url  \
0    Wokingham, Berkshire  https://www.gumtree.com/p/mirrors-clocks-ornam...   
1     Inverness, Highland  https://www.gumtree.com/p/other-household-good...   
2  Skegness, Lincolnshire  https://www.gumtree.com/p/outdoor-settings-fur...   
3      Radstock, Somerset  https://www.gumtree.com/p/other-dining-living-...   
4                    None                                               None   

      page_id create_time  
0  1426704584  2022-02-26  
1  1426704579  2022-02-26  
2  1426704576  2022-02-26  
3  1426704575  2022-02-26  
4        None        None

Now we know a dataframe can be created, it's worth checking that all the rows of data have been imported.

In [9]:
print(len(rows), len(df))

11318 11318


Excellent, all the rows of data have been saved into the dataframe. It's time now to look at the data and see what might need cleaning up.

In [10]:
df.describe()

id     product_name  \
count   11318    8001             8001   
unique  10651    7311             7513   
top            £10.00  Croydon, London   
freq      123      75               20   

                                               category  \
count                                              8001   
unique                                             1280   
top     Other Goods / Hobbies, Interests & Collectibles   
freq                                                286   

                                      product_description       price  \
count                                                8001        7156   
unique                                               7770         439   
top     Specialists in graded & reuse appliances \n Al...  2022-02-27   
freq                                                   30         604   

                      location  \
count                     6311   
unique                    1539   
top     Southampton, Hampshire   
freq                       141   

                                                      url     page_id  \
count                                                6311        6311   
unique                                               6311        6311   
top     https://www.gumtree.com/p/mirrors-clocks-ornam...  1426704584   
freq                                                    1           1   

       create_time  
count         6311  
unique           3  
top     2022-02-27  
freq          4591

Looking at some of the descriptive data, it looks like the problems getting the data straight from the csv and into the dataframe may be due to the missing values(??) There seem to be 11318 rows of data, but only 8001 product rows and various other missing data. It also seems that there are other posts that are missing some of the further information too.

Let's look at the rows where there is no id to see what's going on there.

In [11]:
null_id = df.loc[df['id'].isnull()]
null_id

id product_name category  \
5                      👉 Factory Sell Price  None         None     None   
6                       👉New Amazing Design  None         None     None   
7                  👉 Shannon Sofa Available  None         None     None   
8      👉2 Seater or 3 Seater Sofa Available  None         None     None   
9                👉3+2 Seater Sofa Available  None         None     None   
...                                     ...   ...          ...      ...   
11290                      Collection Only   None         None     None   
11296                       Good condition   None         None     None   
11297                          Woking well   None         None     None   
11298                     Comes with games   None         None     None   
11299                                  £100  None         None     None   

      product_description price location   url page_id create_time  
5                    None  None     None  None    None        None  
6                    None  None     None  None    None        None  
7                    None  None     None  None    None        None  
8                    None  None     None  None    None        None  
9                    None  None     None  None    None        None  
...                   ...   ...      ...   ...     ...         ...  
11290                None  None     None  None    None        None  
11296                None  None     None  None    None        None  
11297                None  None     None  None    None        None  
11298                None  None     None  None    None        None  
11299                None  None     None  None    None        None  

[3317 rows x 10 columns]

Looking at the data more carefully, something that was not picked up in the initial look at df.head(), is that there is a little extra information in the first column that was previously seen to just mimic the index. Here it seems that the first column (in need of a title!) is a description of sorts, most likely provided by the seller. Tere are many products without an id, name or category (looking at this subset of the df and the output of df.describe()) and none of them seem to have any other ifnormation either, except that first column. It's worth confirming what's happening.

In [12]:
null_id.describe()

id product_name  \
count                                                3317    0            0   
unique                                               2931    0            0   
top     - Posting only Mon-Friday [Same day if paid be...  NaN          NaN   
freq                                                   14  NaN          NaN   

       category product_description price location  url page_id create_time  
count         0                   0     0        0    0       0           0  
unique        0                   0     0        0    0       0           0  
top         NaN                 NaN   NaN      NaN  NaN     NaN         NaN  
freq        NaN                 NaN   NaN      NaN  NaN     NaN         NaN

The columns with no id also all seem to have no other information (excluding that first column). It is likely that these are posts that have started to be created by posters/sellers but were never completed.

This was a good inital oversight of the data, but it's also important to check the datatypes - sometimes things aren't as they seem.

In [13]:
df.dtypes

                       object
id                     object
product_name           object
category               object
product_description    object
price                  object
location               object
url                    object
page_id                object
create_time            object
dtype: object

The type of data for the [unnamed],id, product_name, product_description, location, url columns seem okay as objects looking at the dataframe there are strings.

However it may be useful to convert the following:
catgory -> category
price -> float (will need to remove '£')
page_id -> integer
create_time -> time

It's good to have taken an initial look at the data to be able to see what the columns are looking like, in terms of content and datatypes so that when the cleaning pipeline is written, it is with awareness of the data.